apps/cowswap-frontend/src/modules/permit/utils/generatePermitHook.ts

```diff
@@ -105,11 +107,43 @@ async function generatePermitHookRaw(params: PermitHookParams): Promise<PermitHo
 
   const callData = await callDataPromise
 
+  const gasLimit = await calculateGasLimit(callData, owner, tokenAddress, provider, !!account)
+
   return {
     target: tokenAddress,
     callData,
-    // TODO: Find a way to calculate the gas for quote calls without requesting user to sign
-    // For now, keep it hard coded
-    gasLimit: '80000',
+    gasLimit,
+  }
+}
+
+async function calculateGasLimit(
+  data: string,
+  from: string,
+  to: string,
+  provider: Web3Provider,
+  isUserAccount: boolean
+): Promise<string> {
+  if (isUserAccount) {
+    // "Real" user account, use the default
+    // If we send a smaller gas limit than what was used in the quote, it won't matter
+    // If we send a bigger gas limit, the order will fail
+
+    // TODO: it might still happen that user's permit will cost more than the static account
+    return DEFAULT_PERMIT_GAS_LIMIT
+  }
+
+  try {
+    // Query the actual gas estimate
+    const actual = await provider.estimateGas({ data, from, to })
+
+    // Add 10% to actual value to account for minor differences with real account
+    const gasLimit = actual.add(actual.div(10))
+
+    // Pick the biggest between estimated and default
+    return gasLimit.gt(DEFAULT_PERMIT_GAS_LIMIT) ? gasLimit.toString() : DEFAULT_PERMIT_GAS_LIMIT
```

Comment:
I imagine this is because of the `if (isUserAccount)` branch, but in general I don't understand why this is needed.

Reply:
> While testing on Goerli, I was getting the estimation of ~40k for the static account while the user's account was close to 75k.

What? Which token?

Reply:
One solution would be to generate a random private key for doing the signature, this way you are guaranteed to not have collisions.

Reply:
> One solution would be to generate a random private key

That's where this private key came from, but I haven't realized I actually approved it earlier before the whole code was setup 🤦 

I'll create a new one now to be safer.

Reply:
Done.